governator-core/src/main/java/com/netflix/governator/SimpleProvisionMetrics.java

```diff
@@ -1,28 +1,39 @@
 package com.netflix.governator;
 
+import com.google.inject.Key;
+
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Stack;
-import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.TimeUnit;
 
 import javax.inject.Singleton;
 
-import com.google.inject.Key;
-
 @Singleton
-public class SimpleProvisionMetrics implements ProvisionMetrics {
-    private final ThreadLocal<Data> context = new ThreadLocal<Data>();
-    private final CopyOnWriteArraySet<Data> threads = new CopyOnWriteArraySet<>();
+public final class SimpleProvisionMetrics implements ProvisionMetrics {
+    private final ConcurrentMap<Long, Data> threads = new ConcurrentHashMap<>();
     
     public static class Data {
```

Comment:
class name is disjoint from the ProvisioningMetrics function it facilitates; akin to 'Stuff' or 'Thing'. Also I would not expect to find methods in a class named 'Data', only 'data'.